src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -693,32 +693,33 @@ public bool TryGetTopLevelChangeTextVersion(out VersionStamp version)
             return treeAndVersion.Version;
         }
 
-        private static readonly ReaderWriterLockSlim s_syntaxTreeToIdMapLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
         private static readonly ConditionalWeakTable<SyntaxTree, DocumentId> s_syntaxTreeToIdMap =
             new ConditionalWeakTable<SyntaxTree, DocumentId>();
 
         private static void BindSyntaxTreeToId(SyntaxTree tree, DocumentId id)
         {
-            using (s_syntaxTreeToIdMapLock.DisposableWrite())
+            if (s_syntaxTreeToIdMap.TryGetValue(tree, out var existingId))
             {
-                if (s_syntaxTreeToIdMap.TryGetValue(tree, out var existingId))
-                {
-                    Contract.ThrowIfFalse(existingId == id);
-                }
-                else
-                {
-                    s_syntaxTreeToIdMap.Add(tree, id);
-                }
+                Contract.ThrowIfFalse(existingId == id);
+                return;
+            }
+
+            // ConditionalWeakTable is a concurrent map, but doesn't offer a TryAdd method. Use try/catch to behave like
```

Comment:
Any reason not to use [GetValue](https://msdn.microsoft.com/en-us/library/dd413870(v=vs.110).aspx)? Last time I did a pattern like this @KirillOsenkov never let me live it down.